newton/solvers/implicit_mpm/solver_implicit_mpm.py

```diff
@@ -0,0 +1,1893 @@
+
+        self.voxel_size = options.voxel_size
+        self.degree = 1 if options.unilateral else 0
+
+        self.grid_padding = options.grid_padding
+        self.dynamic_grid = options.dynamic_grid
+        self.coloring = options.gauss_seidel
+
+        # Elastic stress-strain matrix from Poisson's ratio and compliance
+        poisson_ratio = options.poisson_ratio
+        lame = 1.0 / (1.0 + poisson_ratio) * np.array([poisson_ratio / (1.0 - 2.0 * poisson_ratio), 0.5])
+        K = options.compliance
+        self.stress_strain_mat = mat66(K / (2.0 * lame[1]) * np.eye(6))
+        self.stress_strain_mat[0, 0] = K / (2.0 * lame[1] + 3.0 * lame[0])
+
+        self._elastic = K != 0.0
+        self._strain_flip_ratio = 0.95
+
+        self.setup_collider(model)
+
+        self.temporary_store = fem.TemporaryStore()
+        fem.set_default_temporary_store(self.temporary_store)
+
+        self._enable_timers = False
+        self._timers_use_nvtx = False
+
+        self._fixed_scratchpad = None
+
+    @staticmethod
+    def enrich_state(state: State):
+        """Enrich the state with additional fields for tracking particle strain and deformation."""
+
+        state.particle_qd_grad = wp.zeros(state.particle_qd.shape[0], dtype=wp.mat33)
+        state.particle_elastic_strain = wp.zeros(state.particle_qd.shape[0], dtype=wp.mat33)
+        state.particle_transform = wp.empty(state.particle_qd.shape[0], dtype=wp.mat33)
+        state.particle_transform.fill_(wp.mat33(np.eye(3)))
+
+        state.velocity_field = None
+        state.impulse_field = None
+        state.stress_field = None
+        state.collider_ids = None
+
+    def setup_collider(
+        self,
+        model: Model,
+        # TODO: read colliders from model
+        colliders: Optional[list[wp.Mesh]] = None,
+        collider_thicknesses: Optional[list[float]] = None,
+        collider_projection_threshold: Optional[list[float]] = None,
+        collider_masses: Optional[list[float]] = None,
+        collider_friction: Optional[list[float]] = None,
+    ):
+        """Setups the collision geometry for the implicit MPM solver.
+
+
+        Args:
+            model: The model to read ground collision properties from.
+            colliders: A list of warp triangular meshes to use as colliders.
+            collider_thicknesses: The thicknesses of the colliders.
+            collider_projection_threshold: The projection threshold for the colliders, i.e, the maximum acceptable penetration depth before projecting particles out.
+            collider_masses: The masses of the colliders.
+            collider_friction: The friction coefficients of the colliders.
+        """
+
+        self._collider_meshes = colliders  # Keep a ref so that meshes are not garbage collected
+
+        if colliders is None:
+            colliders = []
+
+        collider = Collider()
+
+        collider.meshes = wp.array([collider.id for collider in colliders], dtype=wp.uint64)
```

Comment:
Same as above, the arrays need to use `model/state.device` and the kernel launches also need to be on the correct device. You could use a `wp.ScopedDevice` to simplify this.